### ISSUE TITLE
why does mprVolume not include transferFunction imageTexture?

### ISSUE THREAD

**Issue by pgruenbacher-TSUS** [x3dom/x3dom#613]
Beginner here, trying to color a slice of CFD fluid data. In the volrenMPR_aorta.xhtml there is the following code:
```
 <Transform>
            <VolumeData id='volume' dimensions='4.0 4.0 4.0'>
              <ImageTextureAtlas containerField='voxels' url='aorta4096.png'
                                 numberOfSlices='96' slicesOverX='10' slicesOverY='10'/>
              <MPRVolumeStyle id='style' positionLine='0.5'>
                <!--ImageTexture containerField='transferFunction' url='media/volume/transfer/transfer.png'/-->
              </MPRVolumeStyle>
            </VolumeData>
          </Transform>
```

How come when I uncomment the ImageTexture there is still no effect on the MPR volume style? When I look at the documentation the MPRvolume does not appear to have a transferFunctionNode parameter.

**Comment by aarbelaiz** [comment]
The _MPRVolumeStyle_ does not support transfer functions. As you say, in the documentation there is no parameter for it. if you need it, you can apply a transfer function to the whole volume with the _OpacityMapVolumeStyle_

```
<VolumeData id='volume' dimensions='4.0 4.0 4.0'>
   <ImageTextureAtlas containerField='voxels' url='aorta4096.png' numberOfSlices='96' slicesOverX='10' slicesOverY='10'/>
   <OpacityMapVolumeStyle lightFactor='1.2' opacityFactor='6.0'>
      <ImageTexture containerField='transferFunction' url='transfer.png'/>
   </OpacityMapVolumeStyle>
</VolumeData>
```

**Comment by pgruenbacher-TSUS** [comment]
@aarbelaiz If I wanted to get the mprVolume functionality after the opacityMapVolumeStyle, how would I go about doing that? I had earlier tried ComposeVolumeStyle but I don't think OpacityMap or MRPVolumeStyle were valid child types (not composable). Something like this:
```
<VolumeData id='volume' dimensions='4.0 4.0 4.0'>
   <ImageTextureAtlas containerField='voxels' url='aorta4096.png' numberOfSlices='96' slicesOverX='10' slicesOverY='10'/>
   <ComposedVolumeStyle>
       <OpacityMapVolumeStyle/>
       <MPRVolumeStyle/>
   </ComposedVolumeStyle>
</VolumeData>
```
Or would I have the opacity styled VolumeData the child of another VolumeData?
If I wanted to color my MPRVolumeStyled volume data, is the only recourse to fork the code and implement a similar shader as OpacityVolumeStyle or to modify the webgl during runtime?
Or am I going about this wrong way and should be implementing a slicing mechanism on the radar-styled or opacity-styled volume data similar to the clipping plane. http://examples.x3dom.org/clipPlane/clipplane.html

**Comment by aarbelaiz** [comment]
The MPRVolumeStyle is not a composable node, so it can not be composed with other nodes.

Unfortunately, there is not a slicing implementation in the X3D volume rendering nodes.

You have the following options:
- The RadarVolumeSyle will allow you to define a clip plane and a transfer function, in a similar way as the clipplane example, take a look:
http://doc.x3dom.org/developer/x3dom/nodeTypes/RadarVolumeStyle.html

- In case your are trying to only visualize an slice, then use the MPRVolumeRendering (no transfer function). Adding the feature of a transfer function for the MPRVolumeRendering could easily be done. Is this what your are looking for?

- Propose and Implement an slicing mechanism like the clipPlane to all the X3DVolumeRendering nodes, this is a requested feature.

**Comment by pgruenbacher-TSUS** [comment]
I think I'll plan on forking and trying to implement the second option since I'm really only trying to visualize a slice, and it seems like I'd have to just implement similar methods on the mprVolume class as the opacityStyle or the radarStyle, though I'll be sure to track the third option since I think  I saw an issue thread for it. 

@aarbelaiz  If you already know the steps involved for implementing the transfer function for mprVolumeStyle I wouldn't mind some direction, or if there's specific lines of code I should look at first. I figured I'd have to look replicating all the ```if (this._cf.transferFunction.node)``` conditionals such as in the Opacitymap node.  https://github.com/x3dom/x3dom/blob/master/src/nodes/VolumeRendering/OpacityMapVolumeStyle.js#L148-L151

**Comment by aarbelaiz** [comment]
Yes, you should take the OpacityMapVolumeStyle as a reference and replicate the code involving the transfer function node parameter. I'll give it try as soon as I can...

**Comment by pgruenbacher-TSUS** [comment]
thanks!

**Comment by mlimper** [comment]
Thanks @aarbelaiz for the quick solution!
@pgruenbacher-TSUS This is now part of the dev version. Could you test if it works for you?

**Comment by pgruenbacher-TSUS** [comment]
before...
![mprvolume](https://cloud.githubusercontent.com/assets/14790101/12751298/6b3bb32e-c98a-11e5-8158-e8b32bc3baf8.gif)

after...
![colormpr](https://cloud.githubusercontent.com/assets/14790101/12751300/6d653be8-c98a-11e5-9c51-caaec5ffb43c.gif)

same code
```
<mprVolumeStyle is
               finalLine="0.0,1.0,0.0"
               originLine="0.0,0.0,0.0"
               positionLine={position} >
                <imageTexture is
                containerField="transferFunction"
                url={transferFunction}/>
              </mprVolumeStyle>
```

**Comment by npolys** [comment]
NICE!!!

On Tue, Feb 2, 2016 at 8:54 AM, pgruenbacher-TSUS <notifications@github.com>
wrote:

> Closed #613 <https://github.com/x3dom/x3dom/issues/613>.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/x3dom/x3dom/issues/613#event-535849672>.
>



-- 
Nicholas F. Polys, Ph.D.

Director of Visual Computing
Virginia Tech Research Computing

Affiliate Professor
Virginia Tech Department of Computer Science

**Comment by pgruenbacher-TSUS** [comment]
clutch work @aarbelaiz
